tests/components/broadlink/test_config_flow.py

```diff
@@ -0,0 +1,748 @@
+            result["flow_id"], {"host": device.host, "timeout": device.timeout},
+        )
+
+    assert result["type"] == "abort"
+    assert result["reason"] == "already_in_progress"
+
+
+async def test_flow_user_mac_already_configured(hass):
+    """Test we do not accept more than one config entry per device.
+
+    We need to abort the flow and update the existing entry.
+    """
+    device = get_device("Living Room")
+    mock_entry = device.get_mock_entry()
+    mock_entry.add_to_hass(hass)
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    device.host = "192.168.1.64"
+    device.timeout = 20
+    mock_api = device.get_mock_api()
+
+    with patch("broadlink.discover", return_value=[mock_api]):
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], {"host": device.host, "timeout": device.timeout},
+        )
+
+    assert result["type"] == "abort"
+    assert result["reason"] == "already_configured"
+
+    assert dict(mock_entry.data) == device.get_entry_data()
+    assert mock_api.auth.call_count == 0
+
+
+async def test_flow_user_invalid_ip_address(hass):
+    """Test we handle an invalid IP address in the user step."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch("broadlink.discover", side_effect=OSError(errno.EINVAL, None)):
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], {"host": "0.0.0.1"},
+        )
+
+    assert result["type"] == "form"
+    assert result["step_id"] == "user"
+    assert result["errors"] == {"base": "invalid_host"}
+
+
+async def test_flow_user_invalid_hostname(hass):
+    """Test we handle an invalid hostname in the user step."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch("broadlink.discover", side_effect=OSError(socket.EAI_NONAME, None)):
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], {"host": "pancakemaster.local"},
+        )
+
+    assert result["type"] == "form"
+    assert result["step_id"] == "user"
+    assert result["errors"] == {"base": "invalid_host"}
+
+
+async def test_flow_user_device_not_found(hass):
+    """Test we handle a device not found in the user step."""
+    device = get_device("Living Room")
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch("broadlink.discover", return_value=[]):
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], {"host": device.host},
+        )
+
+    assert result["type"] == "form"
+    assert result["step_id"] == "user"
+    assert result["errors"] == {"base": "cannot_connect"}
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6042836)